File: src/main/java/org/noxet/noxetserver/ConfigManager.java
package org.noxet.noxetserver;

import org.bukkit.configuration.file.YamlConfiguration;

import java.io.File;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public abstract class ConfigManager {
    protected YamlConfiguration config;
    private static final Map<String, YamlConfiguration> cache = new HashMap<>();

    protected abstract String getFileName();
    protected String getFileNameWithExtension() {
        return getFileName() + ".yml";
    }

    protected File getFile() {
        File configFile = new File(NoxetServer.getPlugin().getPluginDirectory(), getFileNameWithExtension());

        try {
            if(!((configFile.exists() && configFile.isFile()) || configFile.createNewFile()))
                throw new RuntimeException("Cannot find/create " + getFileNameWithExtension() + " file.");
        } catch(IOException e) {
            throw new RuntimeException(e);
        }

        return configFile;
    }

    protected YamlConfiguration getUncachedConfig() {
        return YamlConfiguration.loadConfiguration(getFile());
    }

    /**
     * Gives the cached configuration if it exists.
     * @return The cached YAML configuration if it exists, otherwise null
     */
    protected YamlConfiguration getCachedConfig() {
        return cache.get(getFileName());
    }

    protected void updateCache() {
        cache.put(getFileName(), getUncachedConfig());
    }

    /**
     * Gives a YAML configuration. Cached if available, otherwise uncached.
     * @return The YAML configuration, cached/uncached
     */
    protected YamlConfiguration getConfig() {
        if(getCachedConfig() == null)
            updateCache();
        return getCachedConfig();
    }

    public ConfigManager() {
        config = getConfig();
    }

    protected void save() {
        try {
            config.save(getFile());
            cache.remove(getFileName());
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        config = getConfig();
    }
}

File: src/main/java/org/noxet/noxetserver/util/ConfigManager.scala
package org.noxet.noxetserver.util

import org.bukkit.configuration.file.YamlConfiguration
import org.noxet.noxetserver.NoxetServer

import java.io.File
import scala.collection.mutable

abstract class ConfigManager:
  protected var config: YamlConfiguration = getConfig

  protected def getFileName: String

  protected def getFileNameWithExtension: String = s"$getFileName.yml"

  protected def getFile: File =
    val configFile = File(NoxetServer.getPlugin.getPluginDirectory, getFileNameWithExtension)

    if !(configFile.isFile || configFile.createNewFile) then
      throw RuntimeException(s"Cannot find/create $getFileNameWithExtension file.")

    configFile

  protected def getUncachedConfig: YamlConfiguration =
    YamlConfiguration.loadConfiguration(getFile)

  protected def getCachedConfig: Option[YamlConfiguration] = ConfigManager.cache.get(getFileName)

  protected def updateCache(): Unit = ConfigManager.cache.put(getFileName, getUncachedConfig)

  /**
   * Gives a YAML configuration. Cached if available, otherwise uncached.
   *
   * @return the YAML configuration, cached/uncached.
   */
  protected def getConfig: YamlConfiguration =
    if getCachedConfig.isEmpty then updateCache()
    getCachedConfig.get

  protected def save(): Unit =
    config.save(getFile)
    ConfigManager.cache.remove(getFileName)
    config = getConfig

object ConfigManager:
  private val cache: mutable.Map[String, YamlConfiguration] = mutable.HashMap()
